### PR TITLE
Add new fog services

### PR DESCRIPTION
**Pull request by netoax**
This PR adds the new fog services to the KNoT custom Linux. The main changes were:

- Add mainflux things service;
- Add mainflux authn service;
- Add babeltower and storage services;
- Bump golang and mainflux versions.

### CODE REVIEW

File: package/knot-storage/knot-storage.mk [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+	mkdir -p $(TARGET_DIR)/usr/local/storage/internal
+	cp -R $(@D)/internal $(TARGET_DIR)/usr/local/storage
+	cp $(KNOT_STORAGE_PKGDIR)/storage.conf $(TARGET_DIR)/usr/local/storage/
+	$(INSTALL) -D -m 0755 $(@D)/bin/cmd $(TARGET_DIR)/usr/local/storage/storage
+	#$(INSTALL) -D -m 0644 $(@D)/internal $(TARGET_DIR)/usr/local/storage
+endef
+
+KNOT_STORAGE_POST_INSTALL_TARGET_HOOKS += KNOT_STORAGE_INSTALL_INIT_SCRIPT
+
+#define KNOT_BABELTOWER_USERS
```
Comment (netoax) [inline package/knot-storage/knot-storage.mk]: Remove this.

File: package/knot-storage/knot-storage.mk [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+define KNOT_STORAGE_INSTALL_INIT_SCRIPT
+	$(INSTALL) -D -m 0755 $(KNOT_STORAGE_PKGDIR)/S105knot-storage $(TARGET_DIR)/etc/init.d/
+endef
+
+define KNOT_STORAGE_INSTALL_TARGET_CMDS
+	mkdir -p $(TARGET_DIR)/usr/local/storage/internal
+	cp -R $(@D)/internal $(TARGET_DIR)/usr/local/storage
+	cp $(KNOT_STORAGE_PKGDIR)/storage.conf $(TARGET_DIR)/usr/local/storage/
+	$(INSTALL) -D -m 0755 $(@D)/bin/cmd $(TARGET_DIR)/usr/local/storage/storage
+	#$(INSTALL) -D -m 0644 $(@D)/internal $(TARGET_DIR)/usr/local/storage
```
Comment (netoax) [inline package/knot-storage/knot-storage.mk]: Remove this.

File: package/knot-babeltower/knot-babeltower.mk [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+define KNOT_BABELTOWER_INSTALL_INIT_SCRIPT
+	$(INSTALL) -D -m 0755 $(KNOT_BABELTOWER_PKGDIR)/S100knot-babeltower $(TARGET_DIR)/etc/init.d/
+endef
+
+define KNOT_BABELTOWER_INSTALL_TARGET_CMDS
+	mkdir -p $(TARGET_DIR)/usr/local/babeltower/internal
+	cp -R $(@D)/internal $(TARGET_DIR)/usr/local/babeltower
+	cp $(KNOT_BABELTOWER_PKGDIR)/babeltower.conf $(TARGET_DIR)/usr/local/babeltower/
+	$(INSTALL) -D -m 0755 $(@D)/bin/cmd $(TARGET_DIR)/usr/local/babeltower/babeltower
+	#$(INSTALL) -D -m 0644 $(@D)/internal $(TARGET_DIR)/usr/local/babeltower
```
Comment (netoax) [inline package/knot-babeltower/knot-babeltower.mk]: Remove this.

File: package/knot-babeltower/knot-babeltower.mk [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+endef
+
+KNOT_BABELTOWER_POST_INSTALL_TARGET_HOOKS += KNOT_BABELTOWER_INSTALL_INIT_SCRIPT
+
+#define KNOT_BABELTOWER_USERS
```
Comment (netoax) [inline package/knot-babeltower/knot-babeltower.mk]: Remove this.

File: package/knot-babeltower/S100knot-babeltower [hunk 5]
```diff
@@ -0,0 +1,34 @@
+NAME=babeltower
+DIR=/usr/local/babeltower
+PID_FILE="/var/run/$NAME.pid"
+
+export `cat ${DIR}/babeltower.conf`
+
+start() {
+        printf "Starting KNoT babeltower Service:"
+	cd $DIR
```
Comment (netoax) [inline package/knot-babeltower/S100knot-babeltower]: Fix indentation.

File: package/go/go.hash [hunk 6]
```diff
@@ -1,2 +1,2 @@
 # From https://golang.org/dl/
-sha256	6264609c6b9cd8ed8e02ca84605d727ce1898d74efa79841660b2e3e985a98bd	go1.10.2.src.tar.gz
+sha256 b13bf04633d4d8cf53226ebeaace8d4d2fd07ae6fa676d0844a688339debec34        go1.13.8.src.tar.gz
```
Comment (netoax) [inline package/go/go.hash]: Adjust the spacing between type and hash.

**Review by netoax — COMMENTED** [review]
Auto-review.

**Comment by netoax** [comment]
I am providing an RPi3 [image](https://drive.google.com/file/d/1YoB-NobsjJn6VCsOowtNM4Yieqt0_ZfC/view?usp=sharing) to whom want to test the changes I added in this PR.

**Comment by netoax** [comment]
There is a known problem when bringing the storage upon the gateway that is related to the database connection settings. In our limited scenario, the storage service starts quicker than the MongoDB daemon and crashes as it can't connect to the database. This fix was already applied but couldn't be added to the tag `KNOT-v03.00-rc01`. However, the storage starts correctly by running `/etc/init.d/S105knot-storage start` explicitly.

Also, another problem that must be fixed is to create a custom RabbitMQ user that can be used by external clients. For testing purposes, you can go through this by reading the [HACKING](https://github.com/CESARBR/knot-gateway-buildroot/blob/master/HACKING.rst) documentation and creating one on hands.

File: package/mainflux/mainflux.mk [hunk 7]
```diff
@@ -4,7 +4,7 @@
 #
 ################################################################################
 
-MAINFLUX_VERSION = master
+MAINFLUX_VERSION = 0.10.0
```
Comment (lcbm) [inline package/mainflux/mainflux.mk:7]: @netoax, don't we need [Mainflux 0.11.0](https://github.com/mainflux/mainflux/releases/tag/v0.11.0), for the [app token (Mainflux's API Key) duration to be configurable to `forever`](https://github.com/CESARBR/knot-cloud-sdk-js-authenticator/pull/17#event-3429943561)? Here's the fixed a [bug](https://github.com/mainflux/mainflux/issues/1103) in 0.11.0.

File: configs/knot_raspberrypi0w_defconfig [hunk 8]
```diff
@@ -57,6 +57,7 @@ BR2_PACKAGE_KNOT_CONTROL=y
 BR2_PACKAGE_KNOT_WEB=y
 BR2_PACKAGE_KNOT_FOG=n
 BR2_PACKAGE_KNOT_FOG_CONNECTOR=y
+BR2_PACKAGE_KNOT_BABELTOWER=y
```
Comment (alla3) [inline configs/knot_raspberrypi0w_defconfig]: I would not add this package on this file.

File: configs/knot_raspberrypi3_defconfig [hunk 9]
```diff
@@ -62,6 +62,7 @@ BR2_PACKAGE_KNOT_CONTROL=y
 BR2_PACKAGE_KNOT_WEB=y
 BR2_PACKAGE_KNOT_FOG=n
 BR2_PACKAGE_KNOT_FOG_CONNECTOR=y
+BR2_PACKAGE_KNOT_BABELTOWER=y
```
Comment (alla3) [inline configs/knot_raspberrypi3_defconfig]: I would not add this package on this file.

File: package/knot-babeltower/Config.in [hunk 10]
```diff
@@ -0,0 +1,12 @@
+config BR2_PACKAGE_KNOT_BABELTOWER
+    bool "knot-babeltower"
+    depends on BR2_PACKAGE_HOST_GO_ARCH_SUPPORTS
+    depends on BR2_PACKAGE_HOST_GO_CGO_LINKING_SUPPORTS
+    depends on BR2_PACKAGE_RABBITMQ_SERVER
+    depends on BR2_PACKAGE_MAINFLUX
```
Comment (alla3) [inline package/knot-babeltower/Config.in:6]: If any of these dependencies are required to build the package, add it to the mk file.

File: configs/knot_raspberrypi0w_defconfig [hunk 11]
```diff
@@ -58,6 +58,7 @@ BR2_PACKAGE_KNOT_WEB=y
 BR2_PACKAGE_KNOT_FOG=n
 BR2_PACKAGE_KNOT_FOG_CONNECTOR=y
 BR2_PACKAGE_KNOT_BABELTOWER=y
+BR2_PACKAGE_KNOT_STORAGE=y
```
Comment (alla3) [inline configs/knot_raspberrypi0w_defconfig]: I would not add this package on this file.

File: configs/knot_raspberrypi3_defconfig [hunk 12]
```diff
@@ -63,6 +63,7 @@ BR2_PACKAGE_KNOT_WEB=y
 BR2_PACKAGE_KNOT_FOG=n
 BR2_PACKAGE_KNOT_FOG_CONNECTOR=y
 BR2_PACKAGE_KNOT_BABELTOWER=y
+BR2_PACKAGE_KNOT_STORAGE=y
```
Comment (alla3) [inline configs/knot_raspberrypi3_defconfig]: I would not add this package on this file.

File: package/knot-storage/Config.in [hunk 13]
```diff
@@ -0,0 +1,10 @@
+config BR2_PACKAGE_KNOT_STORAGE
+    bool "knot-storage"
+    depends on BR2_PACKAGE_HOST_GO_ARCH_SUPPORTS
+    depends on BR2_PACKAGE_HOST_GO_CGO_LINKING_SUPPORTS
+    depends on BR2_PACKAGE_RABBITMQ_SERVER
+    depends on BR2_PACKAGE_MONGODB
+    depends on BR2_PACKAGE_KNOT_BABELTOWER
```
Comment (alla3) [inline package/knot-storage/Config.in:7]: If any of these dependencies are required to build the package, add it to the mk file.

File: package/knot-babeltower/Config.in (hunk 10 above)
Comment (netoax) [reply package/knot-babeltower/Config.in:6]: As we discussed, only libraries need to be added as dependencies in the `.mk` file.

File: package/knot-storage/Config.in (hunk 13 above)
Comment (netoax) [reply package/knot-storage/Config.in:7]: As we discussed, only libraries need to be added as dependencies in the `.mk` file.

File: configs/knot_raspberrypi3_defconfig (hunk 12 above)
Comment (netoax) [reply configs/knot_raspberrypi3_defconfig]: Done.

File: configs/knot_raspberrypi3_defconfig (hunk 9 above)
Comment (netoax) [reply configs/knot_raspberrypi3_defconfig]: Done.

File: configs/knot_raspberrypi0w_defconfig (hunk 8 above)
Comment (netoax) [reply configs/knot_raspberrypi0w_defconfig]: Done.

**Review by alla3 — APPROVED** [review]

**Pull request merged**